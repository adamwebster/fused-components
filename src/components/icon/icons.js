export const icons = {
  times:
    "M24 18.343L42.343 0 48 5.657 29.657 24 48 42.343 42.343 48 24 29.657 5.657 48 0 42.343 18.343 24 0 5.657 5.657 0 24 18.343z",
  "exclamation-circle":
    "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm4,32l-8,0l0,8l8,0l0,-8Zm0,-24l-8,0l0,20l8,0l0,-20Z",
  "question-circle":
    "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm2,32l-0.827,0l-7.173,0l0,8l8,0l0,-8Zm0,-2l0,-2.5c8.915,0.447 11.351,-4.185 11.332,-9.313c-0.037,-9.955 -11.565,-10.182 -11.696,-10.168c0,0 -7.636,-0.019 -7.636,-0.019l0,7l8.207,0c0.015,-0.001 4.004,-0.322 4.004,2.992c0,3.321 -3.97,3.014 -4.035,3.008l-8.176,0l0,9l8,0Z",
  "check-circle":
    "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm-4.173,28.075l17.261,-16.676l5.559,5.754l-23.014,22.234l-0.001,-0.001l-0.001,0.001l-12.837,-13.287l5.754,-5.559l7.279,7.534Z",
  "no-entry-circle":
    "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm0.215,8.001c9.003,0.171 16.932,8.999 15.638,18.235c-1.058,7.549 -8.185,13.764 -15.845,13.764c-10.662,0 -19.612,-12.816 -14.249,-23.282c2.636,-5.143 8.163,-8.605 14.042,-8.717c0.207,-0.001 0.207,-0.001 0.414,0Zm-9.697,10.073c-2.027,3.179 -2.671,7.245 -0.901,10.917c2.795,5.797 10.986,8.026 16.537,4.719l-15.636,-15.636Zm19.168,12.041c3.748,-6.614 0.068,-17.457 -9.528,-17.64c-0.15,-0.001 -0.15,-0.001 -0.3,0c-2.119,0.041 -4.13,0.769 -5.851,1.96l15.679,15.68Z",
  user:
    "M38.59,24c5.72,5.525 9.41,6.232 9.41,16c0,0 0,0 0,0c-14.226,10.656 -33.774,10.656 -48,0c0,0 0,0 0,0c0,-9.768 3.69,-10.475 9.41,-16c2.509,7.466 8.101,8 14.59,8c6.489,0 12.081,-0.534 14.59,-8Zm-14.59,-24c7.682,0 13.919,6.237 13.919,13.919c0,7.682 -6.237,13.919 -13.919,13.919c-7.682,0 -13.919,-6.237 -13.919,-13.919c0,-7.682 6.237,-13.919 13.919,-13.919Z",
  "times-circle":
    "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm0,18.335l10.343,-10.343l5.657,5.657l-10.343,10.343l10.351,10.351l-5.657,5.657l-10.351,-10.351l-10.343,10.343l-5.657,-5.657l10.343,-10.343l-10.335,-10.335l5.657,-5.657l10.335,10.335Z",
  "search":
  "M31.027,36.684c-3.162,2.095 -6.953,3.316 -11.027,3.316c-11.038,0 -20,-8.962 -20,-20c0,-11.038 8.962,-20 20,-20c11.038,0 20,8.962 20,20c0,4.074 -1.221,7.865 -3.316,11.027l11.316,11.316l-5.657,5.657l-11.316,-11.316Zm-11.027,-28.684c6.623,0 12,5.377 12,12c0,6.623 -5.377,12 -12,12c-6.623,0 -12,-5.377 -12,-12c0,-6.623 5.377,-12 12,-12Z",
  "radio": 
  "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm0,8c8.831,0 16,7.169 16,16c0,8.831 -7.169,16 -16,16c-8.831,0 -16,-7.169 -16,-16c0,-8.831 7.169,-16 16,-16Z",
  "radio-checked":
  "M24,0c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24Zm0,8c8.831,0 16,7.169 16,16c0,8.831 -7.169,16 -16,16c-8.831,0 -16,-7.169 -16,-16c0,-8.831 7.169,-16 16,-16Zm0,8c4.415,0 8,3.585 8,8c0,4.415 -3.585,8 -8,8c-4.415,0 -8,-3.585 -8,-8c0,-4.415 3.585,-8 8,-8Z",
  "checkbox":
  "M36,0c6.623,0 12,5.377 12,12l0,24c0,6.623 -5.377,12 -12,12l-24,0c-6.623,0 -12,-5.377 -12,-12l0,-24c0,-6.623 5.377,-12 12,-12l24,0Zm-4,8l-16,0c-4.415,0 -8,3.585 -8,8l0,16c0,4.415 3.585,8 8,8l16,0c4.415,0 8,-3.585 8,-8l0,-16c0,-4.415 -3.585,-8 -8,-8Z",
  "checkbox-checked":
  "M36,48l-24,0c-6.623,0 -12,-5.377 -12,-12l0,-24c0,-6.623 5.377,-12 12,-12l24,0c6.623,0 12,5.377 12,12l0,24c0,6.623 -5.377,12 -12,12Zm-4,-40l-16,0c-4.416,0 -8,3.585 -8,8l0,16c0,4.415 3.585,8 8,8l16,0l0.052,0c4.391,-0.028 7.948,-3.602 7.948,-8l0,-16c0,-4.415 -3.584,-8 -8,-8Zm-11.473,17.904l12.25,-11.835l3.945,4.083l-16.332,15.779l-0.001,-0.001l-0.001,0.001l-9.11,-9.43l4.083,-3.944l5.166,5.347Z",
 "info-circle":
 "M24,48c-13.246,0 -24,-10.754 -24,-24c0,-13.246 10.754,-24 24,-24c13.246,0 24,10.754 24,24c0,13.246 -10.754,24 -24,24Zm-4,-32l8,0l0,-8l-8,0l0,8Zm0,24l8,0l0,-20l-8,0l0,20Z",
};
