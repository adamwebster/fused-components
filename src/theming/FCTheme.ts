import React from 'react';

export interface Theme {
    theme?: string,
    children?: any,
}

export const FCTheme = React.createContext<Theme | null>(null);

export const FCThemeProvider = FCTheme.Provider;

export const FCThemeConsumer = FCTheme.Consumer;