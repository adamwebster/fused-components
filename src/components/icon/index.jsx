import React from 'react';
import { icons } from './icons';

const Icon = ({icon, color}) => {
  return (
    <svg
    xmlns="http://www.w3.org/2000/svg"
    fillRule="evenodd"
    strokeLinejoin="round"
    strokeMiterlimit="2"
    clipRule="evenodd"
    viewBox="0 0 48 48"
  >
    <path
      fill={color || "currentColor"}
      d={icons[icon]}
    ></path>
  </svg>
  )
}

export default Icon;