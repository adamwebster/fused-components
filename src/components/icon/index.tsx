import React from "react";
import { icons } from "./icons";

export interface IconProps {
  icon?: string;
  color?: string;
}
export const Icon = ({ icon, color }: IconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      fillRule="evenodd"
      strokeLinejoin="round"
      strokeMiterlimit="2"
      clipRule="evenodd"
      viewBox="0 0 48 48"
    >
      <path fill={color || "currentColor"} d={icon && icons[icon]}></path>
    </svg>
  );
};
