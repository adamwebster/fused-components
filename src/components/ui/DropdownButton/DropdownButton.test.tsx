import React from 'react';
import { render, cleanup, fireEvent, waitFor, screen } from '@testing-library/react';
import 'jest-styled-components';
import { DropdownButton } from './index';
import { Icon } from '../../icon';
import { FCThemeProvider } from '../../../theming/FCTheme';
import { act } from 'react-dom/test-utils';

afterEach(cleanup);

describe('Dropdown Button Tests', () => {
  test('Renders the Dropdown Button component', () => {
    const { getByText } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    expect(getByText('Dropdown Button')).toBeInTheDocument();
  });

  test('To match snapshot', () => {
    const { getByText } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.click(button);
    expect(getByText('Dropdown Button')).toMatchSnapshot();
  });

  test('To match snapshot dark mode', () => {
    const { getByText } = render(
      <FCThemeProvider value={{ theme: 'dark' }}>
        <DropdownButton id="dm1" label="Dropdown Button">
          <DropdownButton.Menu>
            <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
          </DropdownButton.Menu>
        </DropdownButton>
      </FCThemeProvider>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.click(button);
    expect(getByText('Dropdown Button')).toMatchSnapshot();
  });

  test('The menu shows when the button is clicked', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.click(button);
    expect(getByRole('menu')).toBeInTheDocument();
  });

  test('The menu shows when the button is activated by pressing enter', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.keyDown(button, { key: 'Enter' });
    expect(getByRole('menu')).toBeInTheDocument();
  });

  test('The menu closes when escape is clicked', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.keyDown(button, { key: 'Enter' });
    const menu = getByRole('menu');
    expect(menu).toBeInTheDocument();
    fireEvent.focus(menu);
    act(() => {
      fireEvent.keyDown(menu, { key: 'Escape' });
    });
    expect(screen.queryByRole('menu')).toBeNull();
  });

  test('Clicking the down arrow changes the selected menu item', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.keyDown(button, { key: 'Enter' });
    const menu = getByRole('menu');
    expect(menu).toBeInTheDocument();
    fireEvent.focus(menu);
    act(() => {
      fireEvent.keyDown(menu, { key: 'ArrowDown' });
    });
    expect(menu.getAttribute('aria-activedescendant')).toBe('dm1_menuitem_1');
  });

  test('Clicking the up arrow changes the selected menu item', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.keyDown(button, { key: 'Enter' });
    const menu = getByRole('menu');
    expect(menu).toBeInTheDocument();
    fireEvent.focus(menu);
    act(() => {
      fireEvent.keyDown(menu, { key: 'ArrowDown' });
    });
    act(() => {
      fireEvent.keyDown(menu, { key: 'ArrowUp' });
    });
    expect(menu.getAttribute('aria-activedescendant')).toBe('dm1_menuitem_0');
  });

  test('The menu shows when the down arrow is clicked', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    act(() => {
      fireEvent.keyDown(button, { key: 'ArrowDown' });
    });
    expect(getByRole('menu')).toBeInTheDocument();
  });

  test('The menu does not show when a not defined key is pressed', () => {
    const { getByText, queryByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    act(() => {
      fireEvent.keyDown(button, { key: 'a' });
    });
    expect(queryByRole('menu')).toBeNull();
  });

  test('The menu shows when the button is activated by pressing enter', () => {
    const { getByText, getByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.keyDown(button, { key: 'Enter' });
    expect(getByRole('menu')).toBeInTheDocument();
  });
  test('Clicking an menu item closes the menu', async () => {
    jest.useFakeTimers();

    const { getByText, queryByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.click(button);
    const item = getByText('Menu Item');

    fireEvent.click(item);
    act(() => {
      jest.advanceTimersByTime(400);
    });
    await waitFor(
      () => {
        const menu = queryByRole('menu');
        expect(menu).toBeFalsy();
      },
      { timeout: 300 },
    );
  });

  test('Clicking the button when the menu is open closes the menu', async () => {
    const { getByText, queryByRole } = render(
      <DropdownButton id="dm1" label="Dropdown Button">
        <DropdownButton.Menu>
          <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    const button = getByText('Dropdown Button');
    fireEvent.click(button);
    fireEvent.click(button);

    await waitFor(
      () => {
        const menu = queryByRole('menu');
        expect(menu).toBeFalsy();
      },
      { timeout: 300 },
    );
  });

  test('Clicking outside the menu closes the menu', async () => {
    const { getByText, queryByRole } = render(
      <>
        <button>Click me</button>
        <DropdownButton id="dm1" label="Dropdown Button">
          <DropdownButton.Menu>
            <DropdownButton.MenuItem>Menu Item</DropdownButton.MenuItem>
          </DropdownButton.Menu>
        </DropdownButton>
      </>,
    );
    const button = getByText('Dropdown Button');
    const outSideButton = getByText('Click me');

    fireEvent.click(button);
    fireEvent.mouseDown(outSideButton);
    await waitFor(
      () => {
        const menu = queryByRole('listbox');
        expect(menu).toBeFalsy();
      },
      { timeout: 300 },
    );
  });

  test('Renders the Dropdown Button component when render as a is set', () => {
    const { getByRole } = render(
      <DropdownButton id="dm1" as="a" label={<Icon icon="times" />}>
        <DropdownButton.Menu>
          <DropdownButton.MenuItem icon="check-circle">Like</DropdownButton.MenuItem>
          <DropdownButton.MenuItem>Unlike</DropdownButton.MenuItem>
        </DropdownButton.Menu>
      </DropdownButton>,
    );
    expect(getByRole('img')).toBeInTheDocument();
  });

  test('Renders correctly with a Divider', () => {
    render(
      <DropdownButton id="dm1" as="a" label={<Icon icon="times" />}>
        <DropdownButton.Menu>
          <DropdownButton.MenuItem icon="check-circle">Like</DropdownButton.MenuItem>
          <DropdownButton.Divider />
        </DropdownButton.Menu>
      </DropdownButton>,
    );
  });
});
