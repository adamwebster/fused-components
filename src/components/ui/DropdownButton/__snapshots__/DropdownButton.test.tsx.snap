// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Button Tests To match snapshot 1`] = `
.c1 {
  width: 16px;
  display: block;
  float: right;
  margin-top: 3px;
  height: 5px;
  margin-left: 5px;
}

.c2 {
  width: 1em;
  height: 1em;
}

.c0 {
  vertical-align: middle;
  padding: 0 10px;
  box-sizing: border-box;
  height: 34px;
  cursor: pointer;
  position: relative;
  border-radius: 5px;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: transparent;
  color: #4799ff;
  border: solid 1px #4799ff;
}

.c0:hover:not(:disabled) {
  color: #fff;
  border-color: #147dff;
  background-color: #4799ff;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c0:active:not(:disabled) {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
}

.c0:disabled {
  border-color: #999999;
  color: #999999;
  -webkit-transition: none;
  transition: none;
  cursor: not-allowed;
}

.c0:disabled .sc-AxheI {
  background-color: #999999;
}

<button
  aria-controls="dm1-menu"
  aria-expanded="true"
  aria-haspopup="true"
  class="c0"
  id="dm1"
  tabindex="0"
>
  Dropdown Button
  <span
    class="c1"
  >
    <svg
      class="c2  caret-up"
      clip-rule="evenodd"
      fill-rule="evenodd"
      role="img"
      stroke-linejoin="round"
      stroke-miterlimit="2"
      viewBox="0 0 48 48"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16,0l16,16l-32,0l16,-16Z"
        fill="currentColor"
      />
    </svg>
  </span>
</button>
`;

exports[`Dropdown Button Tests To match snapshot dark mode 1`] = `
.c1 {
  width: 16px;
  display: block;
  float: right;
  margin-top: 3px;
  height: 5px;
  margin-left: 5px;
}

.c2 {
  width: 1em;
  height: 1em;
}

.c0 {
  vertical-align: middle;
  padding: 0 10px;
  box-sizing: border-box;
  height: 34px;
  cursor: pointer;
  position: relative;
  border-radius: 5px;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  background-color: transparent;
  color: #4799ff;
  border: solid 1px #4799ff;
}

.c0:hover:not(:disabled) {
  color: #fff;
  border-color: #147dff;
  background-color: #4799ff;
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.c0:active:not(:disabled) {
  -webkit-transform: scale(0.95);
  -ms-transform: scale(0.95);
  transform: scale(0.95);
}

.c0:disabled {
  border-color: #999999;
  color: #999999;
  -webkit-transition: none;
  transition: none;
  cursor: not-allowed;
}

.c0:disabled .sc-AxheI {
  background-color: #999999;
}

<button
  aria-controls="dm1-menu"
  aria-expanded="true"
  aria-haspopup="true"
  class="c0"
  id="dm1"
  tabindex="0"
>
  Dropdown Button
  <span
    class="c1"
  >
    <svg
      class="c2  caret-up"
      clip-rule="evenodd"
      fill-rule="evenodd"
      role="img"
      stroke-linejoin="round"
      stroke-miterlimit="2"
      viewBox="0 0 48 48"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16,0l16,16l-32,0l16,-16Z"
        fill="currentColor"
      />
    </svg>
  </span>
</button>
`;
